### PR TITLE
fix: Adds fallback dictionary when serialising device info.

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Previously the device tab was serialised using the dictionary supplied from the report, which could be null if the keypath did not exist. This could result in the device tab being completely null. This fix uses an empty dictionary as a fallback so that at least some information (freeDisk, locale, simulator) will always be returned.

### Changed
Initialise empty dictionary and defensively add items from report to it.

## Tests

Added a failing unit test for the encountered scenario then passed it.

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [x] Consistency across platforms for structures or concepts added or modified
- [x] Consistency between the changeset and the goal stated above
- [x] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [x] Usage friction - is the proposed change in usage cumbersome or complicated?
- [x] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [x] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [x] Thoroughness of added tests and any missing edge cases
- [x] Idiomatic use of the language

### CODE REVIEW

**Comment by kastiglione** [comment]
👍

File: Source/BugsnagKSCrashSysInfoParser.m [hunk 1]
```diff
@@ -16,9 +16,10 @@
 #define PLATFORM_WORD_SIZE sizeof(void*)*8
 
 NSDictionary *BSGParseDevice(NSDictionary *report) {
-    NSMutableDictionary *device =
-    [[report valueForKeyPath:@"user.state.deviceState"] mutableCopy];
-    
+    NSMutableDictionary *device = [NSMutableDictionary new];
+    NSDictionary *state = [[report valueForKeyPath:@"user.state.deviceState"] mutableCopy];
```
[the remark this answers is not in the document]
Comment (kastiglione) [reply Source/BugsnagKSCrashSysInfoParser.m]: For consideration: Since `valueForKey:` can throw exceptions for incorrect keys, and since keypaths can't be checked by the compiler, is it worth considering wrapping these in a `@try`/`@catch`? It's certainly very unlikely these will throw, but not impossible.

File: iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+- (void)testEmptyDictSerialisation {
+    // ensures that an empty dictionary parameter returns a fallback dictionary populated with at least some information
+    NSDictionary *device = BSGParseDevice(@{});
+    XCTAssertNotNil(device);
+    XCTAssertNotNil(device[@"locale"]);
+    XCTAssertNotNil(device[@"freeDisk"]);
+    XCTAssertNotNil(device[@"simulator"]);
+}
+
+@end
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m]: I've tested this out and a dictionary is returned with the same contents as if an empty dictionary had been passed in. Adding this as a test would currently fail the suite with an error, presumably due to the fact we've enabled strict build settings and this method parameter is enabled as NonNull. Do you feel we need to change anything so that this can be unit tested?

File: Source/BugsnagKSCrashSysInfoParser.m [hunk 3]
```diff
@@ -16,9 +16,10 @@
 #define PLATFORM_WORD_SIZE sizeof(void*)*8
 
 NSDictionary *BSGParseDevice(NSDictionary *report) {
-    NSMutableDictionary *device =
-    [[report valueForKeyPath:@"user.state.deviceState"] mutableCopy];
-    
+    NSMutableDictionary *device = [NSMutableDictionary new];
```
Comment (martin308) [inline Source/BugsnagKSCrashSysInfoParser.m]: Would it be better to do something like this? I'm not sure how `initWithDictionary` handles nil

```
NSDictionary *state = [report valueForKeyPath:@"user.state.deviceState"];
NSMutableDictionary* device = [[NSMutableDictionary alloc] initWithDictionary: state];
```

File: iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (martin308) [reply iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m]: Is it possible to loosen the restrictions on just the test files? It seems like it might be useful to have a test like that

File: Source/BugsnagKSCrashSysInfoParser.m (hunk 3 above)
Comment (fractalwrench) [reply Source/BugsnagKSCrashSysInfoParser.m]: `initWithDictionary:` requires a Nonnull parameter, whereas `addEntriesFromDictionary:` handles a nil param.

**Comment by fractalwrench** [comment]
@martin308 I've added a test for a nil dictionary by using `#pragma` to loosen the restrictions.

File: iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    XCTAssertNotNil(device[@"locale"]);
+    XCTAssertNotNil(device[@"freeDisk"]);
+    XCTAssertNotNil(device[@"simulator"]);
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wnonnull"
+
+- (void)testNilDictSerialisation {
+    XCTAssertNotNil(BSGParseDevice(nil));
```
Comment (kastiglione) [inline iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m]: This could return an empty dictionary and still pass. I think this should check the contents too.

**Comment by fractalwrench** [comment]
Thanks for the review comments - I've addressed the pragma + validation comments.

File: iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+- (void)validateDeviceDict:(NSDictionary *)device {
+    XCTAssertNotNil(device);
+    XCTAssertNotNil(device[@"locale"]);
+    XCTAssertNotNil(device[@"freeDisk"]);
+    XCTAssertNotNil(device[@"simulator"]);
+}
+
+
+@end
```
Comment (martin308) [inline iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m]: Do we need a test for a valid dictionary that contains the expected key?

Comment (fractalwrench) [reply iOS/BugsnagTests/BugsnagKSCrashSysInfoParserTest.m]: This is tested elsewhere I believe
